edx/edx-enterprise
# ENT-1730 Adding feature specific permission checks to edx-enterprise endpoints

**Description:** 
Remaining work:
- [x] make sure we check that the user's enterprise matches with the request's in the predicates
- [x] map the system wide roles to the feature specific role (just dashboard access for now)
- [x] apply each type of permission to the applicable endpoint in the api views

**JIRA:** https://openedx.atlassian.net/browse/ENT-1730

**Dependencies:** https://github.com/edx/edx-rbac/pull/14, https://github.com/edx/edx-rbac/pull/17

**Testing instructions:**
Waffle switch off:
- [x] a staff user in the enterprise_data_api_access group can hit the with_access_to endpoint and see all available enterprises.
- [x]  an enterprise user in the enterprise_data_api_access group can hit the with_access_to endpoint and see the enterprise they are linked to.
- [x]  a staff user with api credentials can hit any catalog related endpoint for any enterprise.
- [x]  an enterprise user with api credentials can hit any catalog related endpoint for their linked enterprise.
- [x]  a staff user in the enterprise_enrollment_api_access group can post enrollments to the enrollment api endpoint for any enterprise.
- [x]  an enterprise user in the enterprise_enrollment_api_access group can post enrollments to the enrollment api endpoint for their linked enterprise.
- [x]  staff users or enterprise users not in the enterprise_data_api_access group will not get any enterprise returned from the with_access_to endpoint.
- [x]  staff users or enterprise users not in the enterprise_enrollment_api_access group will not be able to post enrollments successfully to the enterprise enrollment api.

Waffle switch on:
We will need to put the following setting in edx-platform/lms/envs/devstack_docker.py:
```
SYSTEM_TO_FEATURE_ROLE_MAPPING = {
    ENTERPRISE_ADMIN_ROLE: [
        ENTERPRISE_DASHBOARD_ADMIN_ROLE
    ],
}
```
There will be a separate ticket to set this setting in production.
Note that this PR does not cover staff role access.
- [x] an enterprise user with the enterprise_admin system wide role will be able to hit the new dashboard-list endpoint and see their linked enterprise returned.
- [x] an enterprise user with the dashboard admin feature specific role will be able to hit the new dashboard-list endpoint and see their linked enterprise returned.
- [x] an enterprise user with the catalog admin feature specific role will be able to hit the catalog related endpoints for their linked enterprise.
- [x] an enterprise user with the enrollment api admin feature specific role will be able to post enrollments to the enrollment api for their linked enterprise.
- [x] enterprise users without any admin roles specified will not have access to the dashboard-list, catalog, or enrollment api endpoints. All other endpoints should behave as before with regards to access.

**Merge checklist:**

  **DO NOT** just add dependencies to `requirements/*.txt` files.

**Post merge:**

**Author concerns:** List any concerns about this PR - inelegant 
solutions, hacks, quick-and-dirty implementations, concerns about 
migrations, etc.
